packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2022 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.internal
+
+import io.realm.RealmInstant
+import io.realm.RealmObject
+import io.realm.dynamic.DynamicMutableRealmObject
+import io.realm.dynamic.DynamicRealmObject
+import io.realm.internal.interop.Link
+import io.realm.internal.interop.RealmValue
+import io.realm.internal.interop.Timestamp
+import io.realm.internal.platform.realmObjectCompanionOrNull
+import kotlin.native.concurrent.SharedImmutable
+import kotlin.reflect.KClass
+
+// Interface for converting storage types (Kotlin representation of Core values) to C-API RealmValue
+internal interface StorageTypeConverter<T> {
+    public fun fromRealmValue(realmValue: RealmValue): T? = realmValueToAny(realmValue) as T?
+    public fun toRealmValue(value: T?): RealmValue = anyToRealmValue(value)
+}
+
+// Top level methods to allow inlining from compiler plugin
+public inline fun realmValueToAny(realmValue: RealmValue): Any? = realmValue.value
+public inline fun anyToRealmValue(value: Any?): RealmValue = RealmValue(value)
+
+// Interface for converting public types to storage types (Kotlin representation of Core values)
+public interface PublicConverter<T, S> {
+    public fun fromPublic(value: T?): S?
+    public fun toPublic(value: S?): T?
+}
+
+// Interface for converting between public type and C-API input/output-values (RealmValue)
+internal interface RealmValueConverter<T> {
```

Comment:
Maybe add a comment why we want to have this alongside the other two convert types?

Reply:
👍 I tried to give an overall brief of the converter-concept/motivation at the top of the file and relate the individual interfaces to that. Hopefully it should be more clear now.